File: ExternalAssets/Scripts/Enemy/GhostController.lua
import 'UnityEngine'

local Life = 3.0;
local MaxOpacity = 1.0;
local MinOpacity = 0.3;
local MaxOpacityDistance = 3.0;
local MinOpacityDistance = 8.0;

local ChaseDistance = 6.0;
local ChaseSpeed = 2.0;
local ChaseMode = false;
local ChargeMode = false;
local ChargeDistance = 1.0;
local ChargeTime = 3.0;
local ExplosionDamage = 1.0;

local IsSpiderTouched = false;

function UpdateOpacity(distanceToPlayer)
    if (distanceToPlayer < MinOpacityDistance) then
        if (distanceToPlayer < MaxOpacityDistance) then
            this.Opacity = MaxOpacity
        else
            this.Opacity = Mathf.InverseLerp(MinOpacityDistance, MaxOpacityDistance, distanceToPlayer)
        end
    end
end

function OnCollisionWithSpider()
    IsSpiderTouched = true
    this.SpiderTouchTimer = aragna.TouchTime
end

function GhostIA()
    local distanceToPlayer = Vector3.Distance(this.transform.position, player.transform.position + Vector3.up)
    UpdateOpacity(distanceToPlayer)

    if (NoMove) then
        return
    end

    if (not ChaseMode and distanceToPlayer < ChaseDistance) then
        ChaseMode = true
    end
    
    if (ChaseMode and not ChargeMode) then
        if (distanceToPlayer > ChargeDistance) then
            ChargeMode = false
            local speed = ChaseSpeed
            if (IsSpiderTouched) then
                speed = speed * aragna.TouchSpeedMultiplier
            end
            this:Move(((player.transform.position + Vector3.up) - this.transform.position).normalized, speed)
        else
            ChargeMode = true
            this.ChargeCountdown = ChargeTime
        end
    end

    if (ChargeMode) then
        if (distanceToPlayer > ChargeDistance) then
            ChargeMode = false
            this.ChargeCountdown = 0.0
        elseif (this.ChargeCountdown == 0.0) then
            player:TakeDamage(ExplosionDamage, this)
            Object.Destroy(this.gameObject)
        end
    end
end

function Start()
    this.Opacity = MinOpacity
    -- AudioManager:Play("Sounds/FANTOME_RODE_07_1")
end

function Update()
    if (this.SpiderTouchTimer == 0.0) then
        IsSpiderTouched = false
    end
    GhostIA()
end

function Die()
    Object.Destroy(this.gameObject)
end

function Knockback(directionFromPlayer)
    -- AudioManager:Play("Sounds/FANTOME_RECOIT_COUP_3_1")
    this:AddForceImpulse(directionFromPlayer * 3.0)
end

function Attacked(power)
    Debug.Log("Attacked")
    Life = Life - power
    NoMove = true
    local directionFromPlayer = (this.transform.position - player.transform.position).normalized
    this:DelayAction(0.32, "Knockback", directionFromPlayer)
    if Life <= 0.0 then
        this:DelayAction(0.7, "Die")
    end
end

File: Assets/Scripts/Gameplay/GhostController.cs
using Cgw.Assets;
using Cgw.Audio;
using Cgw.Scripting;
using System;
using System.Collections;
using System.Net.NetworkInformation;
using UnityEngine;
using UnityEngine.SocialPlatforms;

namespace Cgw.Gameplay
{
    public class GhostController : Enemy
    {
        public float Life = 3.0f;

        public float ChaseDistance = 6.0f;
        public float ChaseSpeed = 2.0f;
        public bool ChaseMode = false;
        public bool ChargeMode = false;
        public float ChargeDistance = 1.0f;
        public float ChargeTime = 3.0f;
        public float ExplosionDamage = 1.0f;

        public bool IsSpiderTouched = false;

        public Color SpriteColor = Color.white;
        public float Opacity = 1.0f;

        public float SpiderTouchTimer = 0.0f;

        public GameObject BoomParticlePrefab;
        public ParticleSystem FlyingDotsParticles;

        private Rigidbody2D m_Rigidbody;
        private SpriteRenderer m_SpriteRenderer;
        private Animator m_Animator;
        private bool NoMove = false;
        private bool Dying = false;

        public void GhostIA()
        {
            if (Dying)
            {
                return;
            }

            Vector3 playerPosition = Player.Instance.transform.position + Vector3.up * 0.4f;
            float distanceToPlayer = Vector3.Distance(transform.position, playerPosition);

            if (NoMove)
            {
                return;
            }

            if (!ChaseMode && distanceToPlayer < ChaseDistance)
            {
                ChaseMode = true;
            }

            if (ChaseMode && !ChargeMode)
            {
                if (distanceToPlayer > ChargeDistance)
                {
                    ChargeMode = false;
                    float speed = ChaseSpeed;
                    if (IsSpiderTouched)
                    {
                        speed *= SpiderController.Instance.TouchSpeedMultiplier;
                    }
                    Move(playerPosition - transform.position, speed);
                }
                else
                {
                    ChargeMode = true;
                }
            }

            if (ChargeMode)
            {
                if (distanceToPlayer > ChargeDistance)
                {
                    ChargeMode = false;
                }
            }
        }

        public void Explode()
        {
            Vector3 playerPosition = Player.Instance.transform.position + Vector3.up * 0.4f;
            float distanceToPlayer = Vector3.Distance(transform.position, playerPosition);
            if (distanceToPlayer < ChargeDistance)
            {
                Player.Instance.TakeDamage(ExplosionDamage, this);
            }
            Instantiate(BoomParticlePrefab, transform.position, Quaternion.identity);
        }

        public void Start()
        {
            m_Rigidbody = GetComponent<Rigidbody2D>();
            m_SpriteRenderer = GetComponentInChildren<SpriteRenderer>();
            m_Animator = GetComponent<Animator>();

            //AudioManager.Instance.Play("Sounds/FANTOME_RODE_07_1");
        }

        public void Update()
        {
            SpiderTouchTimer -= Time.deltaTime;
            SpiderTouchTimer = Mathf.Max(SpiderTouchTimer, 0.0f);

            if (Mathf.Approximately(SpiderTouchTimer, 0.0f) && IsSpiderTouched)
            {
                var emission = FlyingDotsParticles.emission;
                emission.enabled = false;
                IsSpiderTouched = false;
            }
            GhostIA();

            SpriteColor.a = Opacity;
            m_SpriteRenderer.color = SpriteColor;
        }

        public void LateUpdate()
        {
            Vector3 playerPosition = Player.Instance.transform.position + Vector3.up * 0.4f;
            float distanceToPlayer = Vector3.Distance(transform.position, playerPosition);

            m_Animator.SetFloat("DistanceToPlayer", Mathf.InverseLerp(ChargeDistance, ChaseDistance * 0.3f, distanceToPlayer));

            m_Animator.SetBool("Dying", Dying);
            m_Animator.SetBool("ChargeMode", ChargeMode);
        }

        public void Die()
        {
            Dying = true;
            ChargeMode = false;
        }

        public IEnumerator KnockBack(Vector2 directionFromPlayer)
        {
            NoMove = true;
            yield return new WaitForSeconds(0.32f);
            m_Rigidbody.AddForce(directionFromPlayer * 3.0f, ForceMode2D.Impulse);
            yield return new WaitForSeconds(1.0f);
            if (Life <= 0.0f)
            {
                Die();
            }
            NoMove = false;
        }

        public void Move(Vector3 direction, float speed)
        {
            transform.Translate(speed * Time.deltaTime * direction.normalized);
        }

        public override void Attacked(float power)
        {
            if (Dying)
            {
                return;
            }

            Life -= power;
            Vector3 directionFromPlayer = (transform.position - Player.Instance.transform.position).normalized;
            CoroutineRunner.StartCoroutine(KnockBack(directionFromPlayer));
        }

        public override void OnCollisionWithPlayer()
        {
        }

        public override void OnCollisionWithSpider()
        {
            if (SpiderController.Instance.HasTarget)
            {
                var emission = FlyingDotsParticles.emission;
                emission.enabled = true;
                IsSpiderTouched = true;
                SpiderTouchTimer = SpiderController.Instance.TouchTime;
            }
        }

        public override void OnCollisionWithDanger()
        {
        }

        public void OnEndDying()
        {
            Destroy(gameObject);
        }
    }
}